dll/win32/shell32/wine/control.c

```diff
@@ -967,65 +960,201 @@ static	void	Control_DoLaunch(CPanel* panel, HWND hWnd, LPCWSTR wszCmd)
+     * - the CPL path
+     * - either a dialog box number preceeded by @, or a dialog box name.
+     * If there was a second unqoted comma, there is another part of the string:
+     * - the rest of the parameters. */
+    else
+    {
+        /* If there was no second unquoted comma in the string, the CPL path ends at thes
+          * null terminator. */
+        if (!pchSecondComma)
+            pchSecondComma = wszCmd + nLen;
+
+        StringCchCopyNW(buffer, nLen + 1, wszCmd, pchFirstComma - wszCmd);
+        StringCchCopyNW(wszDialogBoxName,
+                        nLen + 1,
+                        pchFirstComma + 1,
+                        pchSecondComma - pchFirstComma - 1);
+
+        if (pchSecondComma != wszCmd + nLen)
         {
-            aCPLName = GlobalAddAtomW(L"CPLName");
+            extraPmts = pchSecondComma + 1;
         }
+    }
+
+    /* Remove the quotes from both buffers. */
+    while ((ptr = StrChrW(buffer, '"')))
+        memmove(ptr, ptr+1, lstrlenW(ptr)*sizeof(WCHAR));
 
-        aCPLFlags = GlobalFindAtomW(L"CPLFlags");
-        if (!aCPLFlags)
+    while ((ptr = StrChrW(wszDialogBoxName, '"')))
+        memmove(ptr, ptr+1, lstrlenW(ptr)*sizeof(WCHAR));
+
+    if (wszDialogBoxName[0] == L'@')
+    {
+        sp = _wtoi(wszDialogBoxName + 1);
+    }
+
+    TRACE("cmd %s, extra %s, sp %d\n", debugstr_w(buffer), debugstr_w(extraPmts), sp);
+
+    applet = Control_LoadApplet(hWnd, buffer, panel);
+    if (applet)
+    {
+        ULONG_PTR cookie;
+        BOOL bActivated;
+        ATOM aCPLName;
+        ATOM aCPLFlags;
+        ATOM aCPLPath;
+        AppDlgFindData findData;
+        
+        /* we've been given a textual parameter (or none at all) */
+        if (sp == -1) 
         {
-            aCPLFlags = GlobalAddAtomW(L"CPLFlags");
+            while ((++sp) != applet->count) 
+            {
+                TRACE("sp %d, name %s\n", sp, debugstr_w(applet->info[sp].name));
+
+                if (StrCmpIW(wszDialogBoxName, applet->info[sp].name) == 0)
+                    break;
+            }
         }
 
-        findData.szAppFile = applet->cmd;
-        findData.sAppletNo = (UINT_PTR)(sp + 1);
-        findData.aCPLName = aCPLName;
-        findData.aCPLFlags = aCPLFlags;
-        findData.hRunDLL = applet->hWnd;
-        findData.hDlgResult = NULL;
-        // Find the dialog of this applet in the first instance.
-        // Note: The simpler functions "FindWindow" or "FindWindowEx" does not find this type of dialogs.
-        EnumWindows(Control_EnumWinProc, (LPARAM)&findData);
-        if (findData.hDlgResult)
+        if (sp >= applet->count && wszDialogBoxName[0] == L'\0')
         {
-            BringWindowToTop(findData.hDlgResult);
+            sp = 0;
         }
-        else
+
+        bActivated = (applet->hActCtx != INVALID_HANDLE_VALUE ? ActivateActCtx(applet->hActCtx, &cookie) : FALSE);
+
+        if (sp < applet->count)
```

Comment:
Replace 'if (sp < applet->count)' with this:
```
if (sp <= applet->count)
```
Needed to fix installation of mshtml on second phase boot.

Reply:
This is not needed if you include the small patch in https://jira.reactos.org/browse/CORE-8981. Thanks.